File: templates/problems/17_math_geometry/05_plus_one/typescript/solution.ts
/**
 * Add one to a number represented as digits. O(n) time, O(1) space (or O(n) if carry propagates).
 */
export function solvePlusOne(digits: number[]): number[] {
    for (let i = digits.length - 1; i >= 0; i--) {
        if (digits[i] < 9) {
            digits[i]++;
            return digits;
        }
        digits[i] = 0;
    }
    // All digits were 9, need extra digit
    const result = new Array(digits.length + 1).fill(0);
    result[0] = 1;
    return result;
}

File: templates/problems/17_math_geometry/05_plus_one/go/solution.go
package plus_one

// solvePlusOne adds one to a number represented as digits. O(n) time, O(1) space (or O(n) if carry propagates).
func solvePlusOne(digits []int) []int {
	for i := len(digits) - 1; i >= 0; i-- {
		if digits[i] < 9 {
			digits[i]++
			return digits
		}
		digits[i] = 0
	}
	// All digits were 9, need extra digit
	result := make([]int, len(digits)+1)
	result[0] = 1
	return result
}
